src/Messages/TSMessagesManager.m

```diff
@@ -281,6 +284,15 @@ - (void)handlePreKeyBundleAsync:(OWSSignalServiceProtosEnvelope *)preKeyEnvelope
 - (void)handleEnvelope:(OWSSignalServiceProtosEnvelope *)envelope plaintextData:(NSData *)plaintextData
 {
     OWSAssert([NSThread isMainThread]);
+
+    BOOL duplicateEnvelope = [self.incomingMessageFinder existsMessageWithTimestamp:envelope.timestamp
+                                                                           sourceId:envelope.source
+                                                                     sourceDeviceId:envelope.sourceDevice];
+    if (duplicateEnvelope) {
```

Comment:
I don't think envelope.sourceDevice will ever be 0, but incomingMessage.sourceDevice will be 0 for models persisted before this code is deployed. Apart from commenting as such in IncomingMessage, do you still think changes are warranted here?